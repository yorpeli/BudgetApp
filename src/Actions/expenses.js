import uuid from 'uuid';

//ADD EXPENSE
export const addExpense = (
    {
        description = '', 
        note = '', 
        amount = 0, 
        createdOn = 0
    } = {}
) => ({
    type: 'ADDEXPENSE',
    expense : {
        id:uuid(),
        description,
        note,
        amount,
        createdOn
    }
});

//REMOVE EXPENSE
export const removeExpense = ({id =0}) => ({
type:'REMOVE',
id
});

//EDIT EXPENSE
export const editExpense = (id, updates) => ({
type:'UPDATE_EXPENSE',
id,
updates
});
