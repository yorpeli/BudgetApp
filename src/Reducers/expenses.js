

const expenseReducerDefaultState =[];

export default (state = expenseReducerDefaultState, action) => {
    switch (action.type){
        case 'ADDEXPENSE':
            return [...state, action.expense];
        case 'REMOVE':
            return state.filter(({id}) => {
                return id !== action.id;
            });
        case 'UPDATE_EXPENSE':
            return state.map((expense)=>{
                if(expense.id === action.id){
                    return{
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

