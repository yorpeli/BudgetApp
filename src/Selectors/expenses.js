import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) =>{
        const createdOnMoment = moment (expense.createdOn);
        const startDaetMatch = startDate ? startDate.isSameOrBefore(createdOnMoment, 'day') : true ; 
        const endDateMatch = endDate ? endDate.isSameOrAfter (createdOnMoment, 'day')  : true;
        const textMatch = typeof text !=='string' || expense.description.toLowerCase().includes(text.toLowerCase());
        return startDaetMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if(sortBy ==='date'){
            return a.createdOn < b.createdOn? 1 : -1;
        } else if (sortBy ==='amount'){
            return a.amount < b.amount? 1: -1;
        }
    });
};