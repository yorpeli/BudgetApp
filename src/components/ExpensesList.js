import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../Selectors/expenses';



export const ExpensesList = (props) => (
    <div>
        {
            props.expenses.length ===0?(
                <p>Nothing to see here</p>
            ):(
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense}/>
                })
            )
        }
    </div>
);


const mapStateToProps = (state) => {
    return{ 
        expenses: SelectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesList);

