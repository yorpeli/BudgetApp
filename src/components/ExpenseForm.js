import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import {addExpenses} from '../Actions/expenses';
import 'react-dates/lib/css/_datepicker.css';

const now = moment();

export default class ExpenseFrom extends React.Component{
    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdOn: props .expense? moment(props.expense.createdOn) : moment(),
            calendarFocused:false,
            error:''
        }
    }
    
    onDescriptionChange = (e) =>{
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) =>{
        const note = e.target.value;
        this.setState (() => ({ note }));
    };
    onAmountChange = (e) =>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,3})?$/)){
            this.setState (() => ({amount}));    
        }
    };

    onDateChange = (createdOn) =>{
        if(createdOn){
            this.setState (() => ({createdOn}));
        }
    };

    onFocusChange = ({ focused }) =>{
        this.setState(() => ({calendarFocused: focused}));
    };

    onSubmit = (e) =>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState (() =>({error: 'Something went Wong!'}));
        } else {
            this.setState(()=>({error:''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdOn: this.state.createdOn.valueOf()
            });
            this.setState(()=> ({
                description:'',
                amount: '',
                note:'',
                createdOn: moment(),
                calendarFocused:false,
                error: ''//`${this.state.description} Submitted Succesfully`
            }));
            
        }
    }
    render(){
        return(
            <div>
                <h3>Expense Form:</h3>
                <form onSubmit = {this.onSubmit}>
                    <input
                        type='text'
                        placeholder = 'description'
                        autoFocus
                        value={this.state.description}
                        onChange = {this.onDescriptionChange}
                    />
                    <input
                        type='number'
                        placeholder= 'amount'
                        value = {this.state.amount}
                        onChange = {this.onAmountChange}    
                    />
                    <SingleDatePicker
                        date = {this.state.createdOn}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {() => false}
                    />
                    <textarea
                        placeholder='add a note to your expense (optional)'
                        value = {this.state.note}
                        onChange = {this.onNoteChange}
                    />
                    <button>Add Expense</button>
                </form>
                { this.state.error? (<p>{this.state.error}</p>):(<p>no errors</p>)}
            </div>
        )
    }
};

