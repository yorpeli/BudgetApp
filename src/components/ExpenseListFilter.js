import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate} from "../Actions/filters";

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused:null
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange = (calendarFocused) => {
        this.setState(()=>({ calendarFocused }))
    };

    render () {
        return(
            <div>
                <input type = "text" 
                value = {this.props.filters.text}  
                onChange = { (e) => {
                    props.dispatch(setTextFilter(e.target.value));
                }}/>
                <select 
                value = {this.props.filters.sortBy}
                onChange = {(e)=>{
                    if(e.target.value === 'amount'){
                        this.props.dispatch(sortByAmount())
                    } else if(e.target.value ==='date'){
                        this.props.dispatch(sortByDate())
                    }
                }}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate}
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange= {this.onFocusChange}
                    showClearDates= {true}
                    numberOfMonths={1}
                    isOutsideRange={()=> false}

                />
            </div>
        )};
}

const mapStateToProps = (state) => {
    return{
        filters:state.filters
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);


