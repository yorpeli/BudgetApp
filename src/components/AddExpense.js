import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../Actions/expenses';
import ExpenseForm from './ExpenseForm';

const AddExpense = (props) => (
    <div>
        <p>Add Expense</p>
        <ExpenseForm
            onSubmit = {(expense)=>{
                console.log(expense);
                props.dispatch(addExpense(expense));
                props.history.push('/');
            }}
        />
    </div>
);

export default connect ()(AddExpense);