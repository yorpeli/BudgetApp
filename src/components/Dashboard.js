import React from 'react';
import ExpensesList from "./ExpensesList"
import ExpenseListFilter from './ExpenseListFilter';

const DashboardPage = () => (
    <div>
        <ExpenseListFilter/>
        <ExpensesList/>
    </div>
);

export default DashboardPage;
