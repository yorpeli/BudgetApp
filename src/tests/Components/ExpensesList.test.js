import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesList } from '../../components/ExpensesList';
import expenses from '../fixtures/expenses';

test ('Render Expense list with Expenses',()=>{
    const wrapper = shallow(<ExpensesList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Should render EMPTY string',()=>{
    const wrapper = shallow(<ExpensesList expenses ={[]}/>);
    expect(wrapper).toMatchSnapshot();
});