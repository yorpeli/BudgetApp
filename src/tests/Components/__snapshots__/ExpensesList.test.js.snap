// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Expense list with Expenses 1`] = `
<div>
  <[object Object]
    amount={7800}
    createdOn={-345600000}
    description="Test Bill"
    id="0"
    key="0"
    note="test desc"
  />
  <[object Object]
    amount={123}
    createdOn={0}
    description="Test Bill2"
    id="1"
    key="1"
    note="test desc"
  />
  <[object Object]
    amount={456}
    createdOn={432000000}
    description="Test credit"
    id="2"
    key="2"
    note="test desc"
  />
</div>
`;

exports[`Should render EMPTY string 1`] = `
<div>
  <p>
    Nothing to see here
  </p>
</div>
`;
